modules/guides/js/guides_in_page_navigation.js

```diff
@@ -0,0 +1,231 @@
+    content.insertBefore(flexWrapper, content.firstChild);
+
+    this.navigation = nav;
+    // Fixed values must be set because of display:fixed when sticky.
+    this.navigation.style.width = flexWrapper.offsetWidth + 'px';
+
+    this.createLookup(headings);
+
+    window.onload = function () {
+      window.addEventListener('scroll', function (e) {
+        self.setSticky();
+        self.setSpy();
+      });
+      var id = window.location.href.split('#')[1];
+      self.scrollTo(id);
+    };
+
+    this.isSetUp = true;
+
+  },
+  createTree: function (headingList) {
+    var self = this;
+    var parent = document.createElement('ul');
+    var result = parent;
+
+    function findParent(node, nodeName) {
+      if (!node) {
+        return null;
+      }
+      else if (node.dataset.parentType === nodeName) {
+        return node;
+      }
+      else {
+        return findParent(node.parentNode, nodeName);
+      }
+    }
+
+    function setParent(i) {
+      // End of the list.
+      if (!headingList[i + 1]) {
+        return;
+      }
+      // The next one is the same type.
+      else if (headingList[i].nodeName === headingList[i + 1].nodeName) {
+        return;
+      }
+      // The next one is a different type.
+      else {
+        var p = findParent(lastInsertedChild, headingList[i + 1].nodeName);
+        if (p !== null) {
+          parent = p;
+        }
+        else {
+          var newParent = document.createElement('ul');
+          var newLi = document.createElement('li');
+          newLi.appendChild(newParent);
+          parent.appendChild(newLi);
+          // Setting the "pointer" to the current head.
+          parent = newParent;
+        }
+      }
+    }
+
+    var lastInsertedChild;
+    var link;
+    var href;
+    for (var i = 0; i < headingList.length; i++) {
+      if (!parent.getAttribute('data-parent-type')) {
+        parent.setAttribute('data-parent-type', headingList[i].nodeName);
+        parent.classList.add('guides__in-page-nav-ul--' + headingList[i].nodeName.toLowerCase());
+      }
+      lastInsertedChild = document.createElement('li');
+      lastInsertedChild.classList.add('guides__in-page-nav-li');
+      href = headingList[i].innerText.toLocaleLowerCase().replace(/\s/g, '-');
+      link = document.createElement('a');
+      link.setAttribute('href', '#' + href);
+      link.setAttribute('title', headingList[i].innerText);
+      link.innerText = headingList[i].innerText;
+      link.addEventListener('click', function (e) {
+        e.preventDefault();
+        var id = this.getAttribute('href').slice(1);
+        self.scrollTo(id);
+      });
+      lastInsertedChild.appendChild(link);
+      parent.appendChild(lastInsertedChild);
+      setParent(i);
+    }
+    return result;
+  },
+  scrollTo: function (id) {
+    if (id) {
+      var scrollToHeading = document.getElementById(id);
+      if (scrollToHeading) {
+        // Subtracting the height of the admin toolbar.
```

Comment:
Maybe should be better to check if is this needed at all (non admin user), measure it dynamically (admin is not on top, default size changed manually etc.)
+ Check it on mobile too.